File: day2b/day2b.jl
#! /usr/local/bin/julia

include("../day2/parse.jl")

majorize(lhs::RGB, rhs::RGB) =
        RGB(max(lhs.red, rhs.red),
                max(lhs.green, rhs.green), max(lhs.blue, rhs.blue))

power(r::RGB) = UInt(r.red) * UInt(r.green) * UInt(r.blue)

function result_b(line::String)
        game, plays = split(line, ':')
        colours = parse.(RGB, split(plays, ';'))
        power(reduce(majorize, colours))
end

function main()
        local file = (length(ARGS) > 0) ? ARGS[1] : "day2/example.txt"
        local sum = 0

        for line = readlines(file)
                sum += result_b(line)
        end
        println("Sum : ", sum)
end

main()

File: day2b/day2b.cpp
#include <cstdlib>
#include <fstream>
#include <istream>
#include <iterator>
#include <numeric>
#include <sstream>
#include <string>

#include "../day2/parse.hpp"

unsigned result_b(std::string line) {
	unsigned id = 0;
	std::string syntax;
	std::stringstream iss{line};
	iss >> syntax >> id >> syntax; // Game %d:

	constexpr auto majorize = [](rgb_t lhs, rgb_t rhs) {
		return rgb_t{std::max(lhs.red, rhs.red),
			     std::max(lhs.green, rhs.green),
			     std::max(lhs.blue, rhs.blue)};
	};

	rgb_t minimal =
	    std::reduce(std::istream_iterator<rgb_t>(iss),
			std::istream_iterator<rgb_t>(), rgb_t{0}, majorize);
	return static_cast<unsigned>(minimal.red) *
	       static_cast<unsigned>(minimal.green) *
	       static_cast<unsigned>(minimal.blue);
}

int main(int argc, const char * argv[]) {
	std::ifstream file{(argc >= 2) ? argv[1] : "day2/example.txt"};
	if (!file) {
		std::cerr << "No file\n";
		return 1;
	}

	size_t sum = 0;
	for (std::string line; std::getline(file, line);) {
		sum += result_b(line);
	}

	std::cout << "Sum : " << sum << '\n';

	return 0;
}
